File: firmware/esp32s3/lib/ble_protocol/ble_protocol.cpp
#include "ble_protocol.h"

namespace ble_protocol {

uint8_t crc8(const uint8_t* data, std::size_t length) {
  uint8_t crc = 0x00;
  for (std::size_t i = 0; i < length; ++i) {
    crc ^= data[i];
    for (int bit = 0; bit < 8; ++bit) {
      if (crc & 0x80) {
        crc = static_cast<uint8_t>((crc << 1) ^ 0x07);
      } else {
        crc = static_cast<uint8_t>(crc << 1);
      }
    }
  }
  return crc;
}

std::size_t encode_fragment(MessageType type,
                            uint16_t sequence_id,
                            uint8_t fragment_index,
                            uint8_t fragment_count,
                            const uint8_t* payload,
                            uint8_t payload_length,
                            uint8_t* out_buffer,
                            std::size_t out_buffer_size) {
  const std::size_t needed = OVERHEAD + payload_length;
  if (out_buffer_size < needed) return 0;
  if (fragment_index >= fragment_count) return 0;

  out_buffer[0] = PROTOCOL_VERSION;
  out_buffer[1] = static_cast<uint8_t>(type);
  out_buffer[2] = static_cast<uint8_t>(sequence_id & 0xFF);         // LE low
  out_buffer[3] = static_cast<uint8_t>((sequence_id >> 8) & 0xFF);  // LE high
  out_buffer[4] = fragment_index;
  out_buffer[5] = fragment_count;
  out_buffer[6] = payload_length;
  for (uint8_t i = 0; i < payload_length; ++i) {
    out_buffer[HEADER_SIZE + i] = payload[i];
  }
  out_buffer[HEADER_SIZE + payload_length] = crc8(out_buffer, HEADER_SIZE + payload_length);
  return needed;
}

DecodeStatus decode_fragment(const uint8_t* data,
                             std::size_t length,
                             DecodedPacket& out_packet) {
  if (length < OVERHEAD) return DecodeStatus::TooShort;
  if (data[0] != PROTOCOL_VERSION) return DecodeStatus::UnknownVersion;

  uint8_t mt = data[1];
  if (mt < 0x01 || mt > 0x04) return DecodeStatus::UnknownMessageType;

  uint8_t payload_length = data[6];
  if (length != static_cast<std::size_t>(OVERHEAD) + payload_length) {
    return DecodeStatus::LengthMismatch;
  }

  uint8_t expected_crc = crc8(data, HEADER_SIZE + payload_length);
  if (expected_crc != data[HEADER_SIZE + payload_length]) {
    return DecodeStatus::CrcMismatch;
  }

  out_packet.header.version         = data[0];
  out_packet.header.message_type    = data[1];
  out_packet.header.sequence_id     = static_cast<uint16_t>(data[2]) |
                                      (static_cast<uint16_t>(data[3]) << 8);
  out_packet.header.fragment_index  = data[4];
  out_packet.header.fragment_count  = data[5];
  out_packet.header.payload_length  = data[6];
  out_packet.payload                = data + HEADER_SIZE;
  out_packet.payload_length         = payload_length;
  return DecodeStatus::Ok;
}

std::vector<FragmentRange> split_utf8(const uint8_t* input,
                                      std::size_t input_length,
                                      std::size_t max_payload) {
  std::vector<FragmentRange> ranges;
  if (max_payload == 0 || input_length == 0) return ranges;

  std::size_t start = 0;
  while (start < input_length) {
    std::size_t end = start + max_payload;
    if (end >= input_length) {
      ranges.push_back({start, input_length - start});
      break;
    }
    // Back off while end points at a UTF-8 continuation byte (10xxxxxx).
    while (end > start && (input[end] & 0xC0) == 0x80) {
      --end;
    }
    if (end == start) {
      // Pathological: a single codepoint longer than max_payload. Emit as-is
      // by advancing to the next leading byte after start.
      end = start + 1;
      while (end < input_length && (input[end] & 0xC0) == 0x80) ++end;
    }
    ranges.push_back({start, end - start});
    start = end;
  }
  return ranges;
}

}  // namespace ble_protocol

File: mobile/lib/ble/ble_protocol.dart
// LingoGlass AR BLE subtitle protocol — Dart mirror of firmware/esp32s3/lib/ble_protocol.
//
// Packet layout is locked in docs/LingoGlass_AR_Development_Plan.md and
// .claude/skills/ble-protocol/SKILL.md. When updating one side, update both.
// Test vectors live in tests/ble_vectors.json.

import 'dart:typed_data';

const int protocolVersion = 0x01;
const int headerSize = 7;
const int trailerSize = 1;
const int packetOverhead = headerSize + trailerSize;
const int maxPayload = 244;

enum MessageType {
  subtitle(0x01),
  status(0x02),
  ack(0x03),
  error(0x04);

  const MessageType(this.value);
  final int value;
}

class PacketHeader {
  const PacketHeader({
    required this.version,
    required this.messageType,
    required this.sequenceId,
    required this.fragmentIndex,
    required this.fragmentCount,
    required this.payloadLength,
  });

  final int version;
  final int messageType;
  final int sequenceId;
  final int fragmentIndex;
  final int fragmentCount;
  final int payloadLength;
}

class DecodedPacket {
  const DecodedPacket(this.header, this.payload);
  final PacketHeader header;
  final Uint8List payload;
}

enum DecodeStatus {
  ok,
  tooShort,
  lengthMismatch,
  crcMismatch,
  unknownVersion,
  unknownMessageType,
}

class DecodeResult {
  const DecodeResult.ok(DecodedPacket this.packet) : status = DecodeStatus.ok;
  const DecodeResult.error(this.status) : packet = null;
  final DecodeStatus status;
  final DecodedPacket? packet;
  bool get isOk => status == DecodeStatus.ok;
}

int crc8(List<int> data, [int start = 0, int? end]) {
  final stop = end ?? data.length;
  int crc = 0x00;
  for (int i = start; i < stop; i++) {
    crc ^= data[i] & 0xFF;
    for (int bit = 0; bit < 8; bit++) {
      if ((crc & 0x80) != 0) {
        crc = ((crc << 1) ^ 0x07) & 0xFF;
      } else {
        crc = (crc << 1) & 0xFF;
      }
    }
  }
  return crc;
}

Uint8List encodeFragment({
  required MessageType type,
  required int sequenceId,
  required int fragmentIndex,
  required int fragmentCount,
  required Uint8List payload,
}) {
  if (payload.length > 0xFF) {
    throw ArgumentError('payload too long: ${payload.length}');
  }
  if (fragmentIndex >= fragmentCount) {
    throw ArgumentError('fragmentIndex >= fragmentCount');
  }
  if (sequenceId < 0 || sequenceId > 0xFFFF) {
    throw ArgumentError('sequenceId out of range');
  }
  final out = Uint8List(packetOverhead + payload.length);
  out[0] = protocolVersion;
  out[1] = type.value;
  out[2] = sequenceId & 0xFF;
  out[3] = (sequenceId >> 8) & 0xFF;
  out[4] = fragmentIndex;
  out[5] = fragmentCount;
  out[6] = payload.length;
  out.setRange(headerSize, headerSize + payload.length, payload);
  out[headerSize + payload.length] = crc8(out, 0, headerSize + payload.length);
  return out;
}

DecodeResult decodeFragment(Uint8List data) {
  if (data.length < packetOverhead) {
    return const DecodeResult.error(DecodeStatus.tooShort);
  }
  if (data[0] != protocolVersion) {
    return const DecodeResult.error(DecodeStatus.unknownVersion);
  }
  final mt = data[1];
  if (mt < 0x01 || mt > 0x04) {
    return const DecodeResult.error(DecodeStatus.unknownMessageType);
  }
  final payloadLength = data[6];
  if (data.length != packetOverhead + payloadLength) {
    return const DecodeResult.error(DecodeStatus.lengthMismatch);
  }
  final expectedCrc = crc8(data, 0, headerSize + payloadLength);
  if (expectedCrc != data[headerSize + payloadLength]) {
    return const DecodeResult.error(DecodeStatus.crcMismatch);
  }
  final header = PacketHeader(
    version: data[0],
    messageType: data[1],
    sequenceId: data[2] | (data[3] << 8),
    fragmentIndex: data[4],
    fragmentCount: data[5],
    payloadLength: payloadLength,
  );
  final payload = Uint8List.sublistView(
    data,
    headerSize,
    headerSize + payloadLength,
  );
  return DecodeResult.ok(DecodedPacket(header, payload));
}

class FragmentRange {
  const FragmentRange(this.offset, this.length);
  final int offset;
  final int length;
}

/// Split UTF-8 bytes into fragments at codepoint boundaries.
/// Each fragment length <= maxPayloadBytes.
List<FragmentRange> splitUtf8(Uint8List input, int maxPayloadBytes) {
  final ranges = <FragmentRange>[];
  if (maxPayloadBytes == 0 || input.isEmpty) return ranges;

  int start = 0;
  while (start < input.length) {
    int end = start + maxPayloadBytes;
    if (end >= input.length) {
      ranges.add(FragmentRange(start, input.length - start));
      break;
    }
    // Back off while end is a UTF-8 continuation byte (10xxxxxx).
    while (end > start && (input[end] & 0xC0) == 0x80) {
      end--;
    }
    if (end == start) {
      // Single codepoint longer than maxPayloadBytes — emit as-is.
      end = start + 1;
      while (end < input.length && (input[end] & 0xC0) == 0x80) {
        end++;
      }
    }
    ranges.add(FragmentRange(start, end - start));
    start = end;
  }
  return ranges;
}
